strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -0,0 +1,374 @@
+            r (float): target squeezing magnitude
+            phi (float): target squeezing phase
+            r_anc (float): squeezing magnitude of the ancillary mode
+            eta_anc(float): detection efficiency of the ancillary mode
+            avg (bool): whether to apply the average or single shot map
+
+        Returns:
+            float: if not avg, the measurement outcome of the ancilla
+        """
+        if avg:
+            self.circuit.mb_squeeze_avg(mode, r, phi, r_anc, eta_anc)
+            return None
+        ancilla_val = self.circuit.mb_squeeze_single_shot(mode, r, phi, r_anc, eta_anc)
+        return ancilla_val
+
+    def beamsplitter(self, theta, phi, mode1, mode2):
+        r"""Implement a beam splitter operation between mode1 and mode2.
+
+        Args:
+            theta (float): real beamsplitter angle
+            phi (float): complex beamsplitter angle
+            mode1 (int): first mode
+            mode2 (int): second mode
+        """
+        self.circuit.beamsplitter(theta, phi, mode1, mode2)
+
+    def gaussian_cptp(self, modes, X, Y=None):
+        r"""Transforms the state according to a deterministic Gaussian CPTP map.
+
+        Args:
+            modes (list): list of modes on which ``(X,Y)`` act
+            X (array): matrix for multiplicative part of transformation
+            Y (array): matrix for additive part of transformation
+        """
+        if isinstance(modes, int):
+            modes = [modes]
+        if Y is not None:
+            X2, Y2 = self.circuit.expandXY(modes, X, Y)
+            self.circuit.apply_channel(X2, Y2)
+        else:
+            X2 = self.circuit.expandS(modes, X)
+            self.circuit.apply_channel(X, Y)
+
+    def measure_homodyne(self, phi, mode, shots=1, select=None, **kwargs):
+        r"""Measure a :ref:`phase space quadrature <homodyne>` of the given mode.
+
+        See :meth:`.BaseBackend.measure_homodyne`.
+        Args:
+            phi (float): angle in phase space for the homodyne
+            mode (int): mode to be measured
+            shots (int): how many samples to collect
+            select (float): if supplied, what value to postselect
+        Keyword Args:
+            eps (float): Homodyne amounts to projection onto a quadrature eigenstate.
+                This eigenstate is approximated by a squeezed state whose variance has been
+                squeezed to the amount ``eps``, :math:`V_\text{meas} = \texttt{eps}^2`.
+                Perfect homodyning is obtained when ``eps`` :math:`\to 0`.
+
+        Returns:
+            array: measured values
+        """
+        # Phi is the rotation of the measurement operator, hence the minus
+        self.circuit.phase_shift(-phi, mode)
+
+        if select is None:
+            val = self.circuit.homodyne(mode, **kwargs)[0, 0]
+        else:
+            val = select * 2 / np.sqrt(2 * self.circuit.hbar)
+            self.circuit.post_select_homodyne(mode, val, **kwargs)
```

Comment:
If you use measure_homodyne, you can either sample a value (using the circuit.homodyne method) or supply the outcome directly (using the circuit.post_select_homodyne). In both cases, however, you also need to update the other modes.

circuit.post_select_generaldyne is the method that performs the update on the other modes.

When you use the circuit.homodyne method, a sample is returned and circuit.post_select_generaldyne is automatically called within circuit.homodyne to update the other modes

When you supply the post-selection value, really all you need is to update the other modes, since you already have the measurement outcome. So when circuit.post_select_homodyne gets called here, it applies circuit.post_select_generaldyne, and the value you passed to it, which is already known, can just directly be used as the returned value of measure_homodyne.